tariff/ostrom.go

```diff
@@ -0,0 +1,266 @@
+		t.cityId, err = t.getCityId()
+		if err != nil {
+			return nil, err
+		}
+		go t.runStatic(done)
+	} else {
+		go t.run(done)
+	}
+	err = <-done
+
+	return t, err
+}
+
+func rate(entry ostrom.ForecastInfo) api.Rate {
+	ts := entry.StartTimestamp.Local()
+	return api.Rate{
+		Start: ts,
+		End:   ts.Add(time.Hour),
+		Price: (entry.Marketprice + entry.AdditionalCost) / 100.0, // Both values include VAT
+	}
+}
+
+func (t *Ostrom) getCityId() (int, error) {
+	var city ostrom.CityId
+
+	params := url.Values{
+		"zip": {t.zip},
+	}
+
+	uri := fmt.Sprintf("%s?%s", ostrom.URI_GET_CITYID, params.Encode())
+	if err := t.GetJSON(uri, &city); err != nil {
+		return 0, err
+	}
+	if len(city) == 0 {
+		// Shouldn't happen, but who knows.
+		return 0, fmt.Errorf("No CityId found for zip: %s", t.zip)
+	} else {
+		return city[0].Id, nil
+	}
+}
+
+func (t *Ostrom) getFixedPrice() (float64, error) {
+	var tariffs ostrom.Tariffs
+
+	params := url.Values{
+		"cityId": {strconv.Itoa(t.cityId)},
+		"usage":  {"1000"},
+	}
+
+	uri := fmt.Sprintf("%s?%s", ostrom.URI_GET_STATIC_PRICE, params.Encode())
+	if err := backoff.Retry(func() error {
+		return backoffPermanentError(t.GetJSON(uri, &tariffs))
+	}, bo()); err != nil {
+		return 0, err
+	}
+
+	for _, tariff := range tariffs.Ostrom {
+		if tariff.ProductCode == ostrom.PRODUCT_BASIC {
+			return tariff.UnitPricePerkWH, nil
+		}
+	}
+
+	return 0, errors.New("Could not find basic tariff in tariff response")
+}
+
+func (t *Ostrom) RefreshToken(_ *oauth2.Token) (*oauth2.Token, error) {
+	tokenURL := ostrom.URI_AUTH + "/oauth2/token"
+	dataReader := strings.NewReader("grant_type=client_credentials")
+
+	req, _ := request.New(http.MethodPost, tokenURL, dataReader, map[string]string{
+		"Authorization": t.basic,
+		"Content-Type":  request.FormContent,
+		"Accept":        request.JSONContent,
+	})
+
+	var res oauth2.Token
+	client := request.NewHelper(t.log)
+	err := client.DoJSON(req, &res)
+
+	if err != nil {
+		t.log.DEBUG.Printf("Requesting token failed with Error: %s\n", err.Error())
+	}
+
+	return util.TokenWithExpiry(&res), err
+}
+
+func (t *Ostrom) GetContracts() ([]ostrom.Contract, error) {
+	var res ostrom.Contracts
+
+	contractsURL := ostrom.URI_API + "/contracts"
+	err := t.GetJSON(contractsURL, &res)
+	return res.Data, err
+}
+
+// This function is used to calculate the prices for the Simplay Fair tarrifs
+// using the price given in the configuration
+// Unfortunately, the API does not allow to query the price for these yet.
+func (t *Ostrom) runStatic(done chan error) {
+	var once sync.Once
+	var val ostrom.ForecastInfo
+	var err error
+	val.AdditionalCost = 0
+
+	tick := time.NewTicker(time.Hour)
+	for ; true; <-tick.C {
+		val.Marketprice, err = t.getFixedPrice()
+		if err == nil {
+			val.StartTimestamp = now.BeginningOfDay()
+			data := make(api.Rates, 0, 48)
+			for i := 0; i < 48; i++ {
+				data = append(data, rate(val))
+				val.StartTimestamp = val.StartTimestamp.Add(time.Hour)
```

Comment:
Wozu braucht es val.StartTimestamp? Das sollte eine lokale Variable sein.

Reply:
Das wird in `rate()` zur  Berechnung von start-und Endzeit verwendet. Ansonsten, siehe mein Kommentar zu "Marketprice"

Reply:
Wie wärs das umzudrehen und die notwendigen Parameter zu übergeben statt sie in ein Struct zu packen?

Reply:
Habe die Funktion jetzt ganz entfernt, da sie nicht mehr nützlich wäre, wenn ich die Parameter sowieso übergebe. Das Ergebnis wird jetzt dierekt in der Schleife berechnet.